### ISSUE TITLE
Patroni in a not so fully distributed world

### ISSUE THREAD

**Issue by fvannee** [zalando/patroni#1593]
This may look like a strange request, as the main feature Patroni advertises itself with is high availability in a distributed world, but I'll try to explain.

Besides HA in a fully distributed manner, Patroni actually has a lot of nice other features. With a single config file, it can initialize masters and create replica's. It is easily customizable using all kinds of hooks in this config file. It has configuration management on a cluster-level, rather than on host-level, using `patronictl`. The command-line tooling `patronictl` has other nice features including a REST API. Basically, it's a nice way to manage your cluster in general, even if we disregard the automatic failover.

However, the way it's currently implemented, automatic failover is an integral and necessary part of Patroni. I like this too though. However, for my use cases, automatic failover (or demotion) triggers too often. It does this, because it wants to guard itself completely against any kind of split-brain scenario. However, in my case, I find an unnecessary demotion (+restart) to be potentially worse than a possible split-brain.

I'm most worried about a scenario like the following:

- There's an ETCD cluster running somewhere and a Postgres server A and B managed by Patroni.
- B is replicating from A
- Now somehow ETCD cluster becomes unavailable (for example due to a network split). A and B can still communicate, but they cannot update their state in ETCD anymore.
- Both A and B notice that ETCD is down. Patroni on B also does a check that A is still alive and finds that that is the case.
- However, in the current scenario, B stops following A. Also, A demotes itself. They do this, because they think there may be a server C somewhere that can still reach ETCD and that promotes itself to the new master. They want to avoid split-brain, so they demote themselves.

In our environment, I'd rather have both A and B do nothing if they detect that they can still communicate with each other, rather than that they demote+restart themselves.
Basically, I'd like to have all the useful features of Patroni as written above, without this added extra risk of ETCD going down / becoming unreachable in the chain.

What are your opinions on adding an option to the Patroni config file to optionally change this behavior? The new behavior would still automatic failover under normal situations where primary suddenly becomes unavailable, but won't do the preventive demoting/stop following of primary/secondary in case ETCD becomes unreachable.

**Comment by CyberDem0n** [comment]
> for my use cases, automatic failover (or demotion) triggers too often.

This is the clear sign that something is wrong with configuration.

> In our environment, I'd rather have both A and B do nothing if they detect that they can still communicate with each other, rather than that they demote+restart themselves.

How these two can make sure that the third node didn't join the cluster and become the leader?

Do you realize that in order to correctly implement the feature you want you'll have to implement the protocol similar to Raft or Paxos?

**Comment by fvannee** [comment]
> This is the clear sign that something is wrong with configuration.

This is just some testing I did with all kinds of edge cases, when I manually bring etcd down for example.

> How these two can make sure that the third node didn't join the cluster and become the leader?

My point is more that for the kind of clusters I'm talking about, I don't care if in that particular situation a third node joins and becomes the leader. In our setups that's so unlikely that I don't care. Most of the setups are rather static. I don't need the completely full automatic failover. I'd like to use Patroni mostly for the ease of use of configuration management and setup that Patroni offers. That's why my question was phrased with 'in a not so distributed world'.

If, in the situation above,a third node happens to join the cluster (and it can somehow reach etcd) and it becomes master. Yes, then there would be two masters and there'd be a split brain. That's a risk I'd be willing to accept.

**Comment by CyberDem0n** [comment]
> when I manually bring etcd down for example.

Well, if you shoot your own foot it hurts.

> Yes, then there would be two masters and there'd be a split brain. That's a risk I'd be willing to accept.

You are not the first one coming here with such suggestions, but my opinion on this topic will not change. I am against having a foot gun in Patroni.

P.S. maybe you don't need Etcd at all, and should try https://github.com/zalando/patroni/pull/375
It would allow running the cluster with two postgres nodes and the third node with just a raft-controller. If only a single component in such setup fails everything will continue to work. I.e. if the raft-controller is down the primary-replica pair will continue working. If the replica is down but raft-controller is working - everything is fine. If primary failed but there is a replica and raft-controller - replica will promote.

**Comment by fvannee** [comment]
> Well, if you shoot your own foot it hurts.
I gotta have some way to test scenarios like a split network ;-)

> P.S. maybe you don't need Etcd at all, and should try #375
I'll take a look at this. That looks promising.

**Comment by fvannee** [comment]
I'll close this issue now. Regarding the PR #375 - what are the plans to merge this to master? I noticed it's been open for a long time already.
